Replace Go-js with Go-wasi

tsahee opened the pull request:
## What's working?
system tests, including stylus, in both jit and arbitrator mode

```
go version #remember to use 1.21

make test-go-deps
go test -tags stylustest -run TestProgram ./system_tests/...

make target/machines/latest/user_test.wasm
cd arbitrator
cargo test
```

## What's not working?
ipfshelper - the library we use is out-of-date. I stubbed out enough of it to be able to run system tests, but things like make build / make lint will not work.

one-step-proofs - not done yet.

## recommended order for reading
### arbcompress/, wavmio/
These changes are required for go 1.21, even if staying with wasm-js

### arbitrator/wasm-libraries/{brotli, host-io}
Equivalent changes in the wasm libs

### arbos/programs
Updated API to execute stylus programs

### arbitrator/prover/src/{machine, wavm}.rs
required modifications for supporting the stylus->go callbacks. OSP not ready here but should be simple considering we only ever need state of either first or last {value, frame} stack

### arbitrator/prover/src/host.rs
funcs using these new opcodes

### arbitrator/wasm-libs/{user-host, program-exec}
implementation from arbitrator side

### arbitrator/jit
the jit implementation

tsahee committed:
arbutil: add SliceToUnsafePointer

tsahee committed:
arbcompress: wasm support go 1.21

tsahee committed:
brotli lib: switch to go-wasi

tsahee committed:
host-io wasm: switch to go-wasi interface

tsahee committed:
wavmio: switch to go 1.21 support

tsahee committed:
wasi stub: implement missing functions for go-wasi

tsahee committed:
makefile changes

tsahee committed:
improve and update go-testcase

tsahee committed:
arbutil: don't compile for wasm what isn't needed

tsahee committed:
machine: add go/wasi entry support

tsahee committed:
prover machine: support cothreads, without proving

tsahee committed:
prover machine: add functions to switch cothreads

tsahee committed:
add program_exec wasm module

tsahee committed:
wasm-libraries: use num_enum, remove old go-abi

tsahee committed:
EvmAPI: initial

tsahee committed:
machine: remove error guard policy

Replace with cothread

tsahee committed:
program: arbitrator support with go-wasi

tsahee committed:
go-wasi: jit support

tsahee committed:
go-wasi: update Makefile

tsahee committed:
arbitrator: cargo fmt

tsahee committed:
Merge remote-tracking branch 'origin/stylus' into go_wasi

tsahee committed:
stylus/prover: remove unused parameter from program_continue

tsahee committed:
go-ethereum: support go-wasi

tsahee committed:
fastcache: support go-wasi

tsahee committed:
TEMP: stub out ipfs service

ipfs support uses unsupported libs that fail when switching to go 1.21.
This stubs them out from das, enabling system_tests to compile.

tsahee committed:
arbutil: rename Js->Requestor

no longer js

tsahee commented on arbitrator/arbutil/src/evm/req.rs:
```diff
@@ -0,0 +1,250 @@
+// Copyright 2023, Offchain Labs, Inc.
+// For license information, see https://github.com/OffchainLabs/nitro/blob/master/LICENSE
```
note: this is based on the old js.rs file, but seems like changed enough that github doesn't recognize it.

tsahee commented on arbitrator/arbutil/src/evm/req.rs:
```diff
@@ -0,0 +1,250 @@
+    }
+}
+
+impl<T: RequestHandler> EvmApi for EvmApiRequestor<T> {
```
Note: this is now the only implementation of EvmApi trait.

tsahee replied:
(except for the test one)

tsahee commented on arbitrator/jit/src/arbcompress.rs:
```diff
@@ -1,7 +1,8 @@
-use crate::{gostack::GoStack, machine::WasmEnvMut};
+use crate::machine::Escape;
+use crate::{goenv::GoEnv, machine::WasmEnvMut};
 
```
note: I am thinking of merging this code with that of wasm-libraries/brotli - but not high prio

tsahee commented on arbitrator/jit/src/goenv.rs:
```diff
@@ -0,0 +1,180 @@
+// Copyright 2022, Offchain Labs, Inc.
+// For license information, see https://github.com/nitro/blob/master/LICENSE
+
```
note: this file is based on the original gostack.rs.
I plan to eventually have this in a trait that will allow sharing more code bewteen jit and wasm-libraries.

tsahee commented on arbitrator/jit/src/program.rs:
```diff
@@ -0,0 +1,235 @@
+// Copyright 2022-2023, Offchain Labs, Inc.
```
note: bthis is based on/replaces previous user submodule

tsahee commented on arbitrator/jit/src/wasip1_stub.rs:
```diff
@@ -0,0 +1,305 @@
+// Copyright 2021-2023, Offchain Labs, Inc.
+// For license information, see https://github.com/nitro/blob/master/LICENSE
```
note: near future plan: merge this code with wasm-libraries/wasi-stub

tsahee commented on arbitrator/prover/src/error_guard.rs:
```diff
@@ -8,30 +8,7 @@ use crate::{
-        &mut self.guards
-    }
-}
+use std::fmt::{self, Display};
```
note: enable/disable error guard is not needed. Instead, we use "cothread" boolean which does the same

tsahee committed:
jit/program: improve documentation

tsahee committed:
prover/machine: remove support for go/js entrypoint

tsahee committed:
wasm-library programs: document

tsahee committed:
arbitrator-test: fix TestEvmApi

tsahee committed:
test-cases: don't improt set_error_policy

anodar committed:
Implement prove_stacks for stack of stacks

anodar committed:
Drop comment

anodar committed:
Implement hashing of all value/frame stacks in hashing the machine

anodar committed:
Reorder stack elements to from 0...n-1 to 0,n-1,1,2,...,n-2 in hashing to match order in proving

anodar committed:
Change order of elements in  hashing/proving

anodar committed:
Reorder multistack hashing, add cothread into proving and hashing

anodar committed:
Address rest of the comments

rachel-bousfield commented on arbos/programs/api.go:
```diff
@@ -17,277 +18,309 @@ import (
+			}
+			cost := vm.WasmStateStoreCost(db, actingAddress, key, value)
+			db.SetState(actingAddress, key, value)
+			return []byte{1}, cost
```
In the previous model, we got to work with nicely typed closures. Sure there was byte-munging, but it was at the type level vs baked into the business logic. Let's look at this function but the reasoning applies across all the hostios.
```go
setBytes32 := func(key, value common.Hash) (uint64, error) {
    if tracingInfo != nil {
        tracingInfo.RecordStorageSet(key, value)
    }
    if readOnly {
        return 0, vm.ErrWriteProtection
    }
    cost := vm.WasmStateStoreCost(db, actingAddress, key, value)
    db.SetState(actingAddress, key, value)
    return cost, nil
}
```

In the new code, Rust and Go have to agree on special bytes and structure. This is a bit dangerous in constructs where the shape is variable length like in the contract calling hostios, but even in this simpler hostio we get an example.

In the old model, we could nicely pass `vm.ErrWriteProtection` back to the Rust side. Here though we have to set a flag byte, and then expect Rust treat it the same way. It turns out that Rust now just says "set_bytes32 failed", but that's not exactly what happened. We hit a write protection, which users can do by static calling into a contract that performs a write.

It's really hard to see this kind of nuance across the Rust <=> Go interop line when all these concepts are happening at the same time. I think we need to find a way to replace this switch statement with something more type-aware. A reader pattern could be a first start, where instead of reading offsets explicitly you read from a stream of bytes one type at a time. We do this in a few places like `TestProgramEvmData`. In terms of making sure the length is right, you could error if the stream runs out early or has extra bytes at the end.

tsahee replied:
as discussed - kept the switch statement for encoding, decoding only

rachel-bousfield commented on arbos/programs/wasm.go:
```diff
@@ -67,20 +60,64 @@ func activateProgram(
 	debug bool,
 	burner burn.Burner,
 ) (*activationInfo, error) {
+	errBuf := make([]byte, 1024)
```
In debug mode these can be very long, I wouldn't use a fixed size

rachel-bousfield commented on arbitrator/arbutil/src/evm/req.rs:
```diff
@@ -0,0 +1,250 @@
+        let mut request = vec![];
+        request.extend_from_slice(contract.as_slice());
+        request.extend_from_slice(value.as_slice());
+        request.extend_from_slice(&gas.to_be_bytes());
```
Here and elsewhere you can do
```rs
request.extend(contract)
```

They're the same just `extend_from_slice` is going to be deprecated and `extend` is prettier :)

rachel-bousfield commented on arbitrator/arbutil/src/evm/req.rs:
```diff
@@ -0,0 +1,250 @@
+
+pub struct EvmApiRequestor<T: RequestHandler> {
+    handler: T,
+    last_call_result: Vec<u8>,
```
In this model, are we able to return data from Go => Rust without the copy?

For example, the `read_return_data` hostio should ideally just pass the pointer from Go back to Rust (Geth takes special care of that slice so that it's zero-copy).

tsahee replied:
will be implemented later

rachel-bousfield commented on arbitrator/arbutil/src/evm/req.rs:
```diff
@@ -0,0 +1,250 @@
+            let mut err_string = String::from("create_response_malformed");
+            if res.len() > 1 {
+                let res = res.drain(1..).collect();
+                match String::from_utf8(res) {
```
```rs
String::from_utf8_lossy(&res[1...])
String::from_utf8(res).unwrap_or("create_response_malformed".into())

// or even
if let Ok(str) = String::from_utf8(res) {
    err_string = str
}
```

rachel-bousfield commented on arbitrator/arbutil/src/evm/req.rs:
```diff
@@ -0,0 +1,250 @@
+        } else {
+            vec![]
+        };
+        return (Ok(address), self.last_call_result.len() as u32, cost);
```
You can drop the `return` here

rachel-bousfield commented on arbitrator/arbutil/src/evm/req.rs:
```diff
@@ -0,0 +1,250 @@
+            );
+        }
+        let address = res.get(1..21).unwrap().try_into().unwrap();
+        self.last_call_result = if res.len() > 21 {
+            res.drain(21..).collect()
+        } else {
+            vec![]
+        };
```
Maybe
```rs
self.last_call_result = (res.len() > 21)
    .then(|| res.drain(21..).collect())
    .unwrap_or_default();
```

tsahee replied:
improved it with pre-draining.

rachel-bousfield commented on arbitrator/arbutil/src/evm/req.rs:
```diff
@@ -0,0 +1,250 @@
+
+impl<T: RequestHandler> EvmApi for EvmApiRequestor<T> {
+    fn get_bytes32(&mut self, key: Bytes32) -> (Bytes32, u64) {
+        let (res, cost) = self
+            .handler
```
`EvmApiRequestor` should probably implement `Deref<RequestHandler>` and `DerefMut<RequestHandler>` to make these kinds of expressions shorter

tsahee replied:
implemented in the next PR by having a handle_request in this func

rachel-bousfield commented on arbitrator/arbutil/src/evm/req.rs:
```diff
@@ -0,0 +1,250 @@
+        if offset + size > data_len {
+            size = data_len - offset;
+        }
+        data[offset..size].to_vec()
```
We should avoid this copy and instead return a slice

tsahee replied:
will be implemented later

rachel-bousfield commented on arbitrator/arbutil/src/evm/req.rs:
```diff
@@ -0,0 +1,250 @@
+        if res.is_empty() {
+            Ok(())
+        } else {
+            Err(eyre!(
```
`bail!()`

Also I like
```rs
match cond {
    true => expr,
    false => expr,
}
```

It's pretty with verum and falsum as ligatures :)

rachel-bousfield commented on arbitrator/jit/src/program.rs:
```diff
@@ -0,0 +1,251 @@
+    err_buf: Uptr,
+    err_buf_len: u32,
+) -> Result<u32, Escape> {
+    let mut genv = GoEnv::new(&mut env);
```
Let's call it `go_env` or `env`

tsahee replied:
renamed class to CallerEnv and var to caller_env..
it's not really go specific now

rachel-bousfield commented on arbitrator/jit/src/stylus_backend.rs:
```diff
@@ -0,0 +1,173 @@
+        if self
+            .tx
+            .send(MessageFromCothread {
+                req_type: req_type as u32 + 0x10000000,
```
We should document `0x10000000`

rachel-bousfield commented on arbitrator/wasm-libraries/user-host/src/program.rs:
```diff
@@ -163,9 +242,13 @@ impl UserHost for Program {
     }
+
+    fn start_ink(&self) -> Result<u64,Self::Err> {
+        return Err(eyre!("recording start ink while proving").into())
```
`bail!`

tsahee replied:
bail here doesn't work because it returns the wrong error (need that into())

rachel-bousfield reviewed (changes requested):
Nice draft!

I need to review more, but wanted to provide some feedback earlier on.

I know this is an early but something that'd help would be docstrings. There's a lot of subtle stuff happening and across multiple files so any kind of documentation in the code or in notion would be helpful.

That and when the repo builds, I'll be able to try more things out. In any case here's a few comments I hope will make this even better :)

anodar committed:
Cargo fmt, fix match statements, drop static constructor from Bytes32

anodar committed:
Extra colon in comment

anodar committed:
Change match to if, inline cothread_to_bytes

tsahee committed:
machine: remove error_guards

they don't work with cothreads because the main thread's stack might shrink after calling a program for the first time (while handeling requests)

anodar committed:
Merge branch 'go_wasi' into multiple_stacks

anodar committed:
Drop commented out fn

tsahee committed:
stylus api: separate encoding/decoding from logic

tsahee committed:
switchthreads takes recovery offset

tsahee committed:
programs/api.go: remove unnecessary diffs

tsahee committed:
Merge branch 'go_wasi' into multiple_stacks

tsahee committed:
Merge remote-tracking branch 'origin/stylus' into go_wasi

tsahee committed:
go-wasi: fix accountCode{,Size}

tsahee committed:
stylus: fix accountCode

tsahee committed:
machine: fix multistack hashing

tsahee committed:
Merge branch 'go_wasi' into multiple_stacks

tsahee committed:
prover: use main_thread_recovery pc

also fix proving around cothread

tsahee committed:
arbitrator: cargo fmt

tsahee committed:
contract: support cothreads

tsahee committed:
arbitrator: rust and naming improvements

tsahee committed:
stub-out ipfs support

tsahee committed:
programs/api: remove inefectual code

tsahee committed:
ipfs: remove unnecessary changes when stubbing

tsahee committed:
update ci to go 1.21

tsahee committed:
use constant instead of magic number for evm_api_offset

tsahee committed:
test constants are identical in rust and go

tsahee committed:
remove js-runtime from ci test

tsahee committed:
fix dockerfile

tsahee committed:
multistack: remaining uses standard stack_hashing

0 if no stacks

tsahee committed:
contracts: fix multistack

tsahee committed:
cargo fmt

tsahee committed:
fix dockerfile

tsahee committed:
update contracts
